### DISCUSSION TITLE
How do I get a JSON value from a `Record`?

### DISCUSSION THREAD

**Discussion by Nv7-GitHub** [pocketbase/pocketbase#547] [Q&A] [answered]
How do I get the value of a JSON field from a `*models.Record` object?

**Comment by ganigeorgiev** [comment, marked as answer]
You can try retrieving the string value of the stored json and unmarshal it into a struct/map:

```go
result := struct {
	firstName string `json:"first_name"`
}{}

err := json.Unmarshal([]byte(record.GetStringDataValue("field")), &result)
```

Alternatively, if you want the json value as raw bytes slice, you can also do something like `types.ParseJsonRaw(record.GetDataValue("field"))`. This will return a `types.JsonRaw` which is just a db encodable alias for `[]byte`.

_I'll add a helper accessor method in the next v0.8.0 release to simplify this into a one-liner._

**Reply by Nv7-GitHub** [reply]
Oh ok, I was doing the first method but was wondering if there was a cleaner method.

**Reply by ganigeorgiev** [reply]
Currently no. With v0.8.0 there will be a `record.UnmarshalJSONField("field", &result)` method (_and other "cosmetic" changes like renaming `GetStringDataValue` to just `GetString`, etc._).

**Reply by YoRHa-A5** [reply]
How would one go about retrieving nested json values? 🤔 

```
"expand": {
  "location": {
    "latitude": 12.34,
    "longitude": 56.78
  }
}
```

**Reply by ganigeorgiev** [reply]
@BunnyTheLifeguard Could you elaborate? From where is coming the above json object and where do you want to retrieve it? As part of the `filter` request param on inside a custom Go route?

**Reply by YoRHa-A5** [reply]
@ganigeorgiev sure thing, it is in a custom hook `app.OnRecordsListRequest("users").Add(func(e *core.RecordsListEvent) error` but maybe it is better to open an own discussion since I don't want to clutter this one and I might need some more info for implementing my custom functionality 😃 #2705

**Reply by ched-dev** [reply]
> How would one go about retrieving nested json values? 🤔
> 
> ```
> "expand": {
>   "location": {
>     "latitude": 12.34,
>     "longitude": 56.78
>   }
> }
> ```

I was looking for an answer to this myself, so here is an example I was able to get working which shows how to access JSON Field Objects and JSON Field Arrays:

```js
  const pbPlaylists = $app.findAllRecords('playlists')

  // unmarshal with only specific props
  let playlists = pbPlaylists.map((pbPlaylist) => {
    const playlist = new DynamicModel({
      playlist_id: pbPlaylist.get('playlist_id'),
      playlist_slug: pbPlaylist.get('playlist_slug'),
      playlist_title: pbPlaylist.get('playlist_title'),
      playlist_description: pbPlaylist.get('playlist_description'),
      playlist_duration: pbPlaylist.get('playlist_duration'),
      playlist_thumbnails: new DynamicModel({
        default: {},
        high: {},
        maxres: {},
        medium: {},
        standard: {}
      }),
      playlist_videos: []
    })
    pbPlaylist.unmarshalJSONField('playlist_thumbnails', playlist.playlist_thumbnails)
    pbPlaylist.unmarshalJSONField('playlist_videos', playlist.playlist_videos)
    return playlist
  })

  // remove hidden fields for anonymous users
  if (playlists && !includeEmbed) {
    playlists = playlists.map(removeProtectedPropertiesFromPlaylist)
  }

  return playlists
}
```

- The values I `.get()` are simple strings passing along the value
- I have a `playlist_thumbnails` JSON field that is an object and needs to be unmarshalled, and it requires a nested `DynamicModel` to unmarshal properly, but only the top level needs to be defined
- I have a `playlist_videos` JSON field that is an array of objects but it only needs an array definition to unmarshal correctly
- Call the `.unmarshalJSONField()` for each of the JSON field you have

In my use case, I needed to modify the JSON field value before returning to the user. This requires unmarshalling to access the JSON values.
